Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/view/CountryList.kt

```diff
@@ -2,47 +2,130 @@ package de.rki.coronawarnapp.ui.view
+import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.RecyclerView
 import de.rki.coronawarnapp.R
+import de.rki.coronawarnapp.databinding.ViewCountryListEntryFlagItemBinding
 import de.rki.coronawarnapp.ui.Country
+import de.rki.coronawarnapp.util.lists.BindableVH
+import kotlinx.android.synthetic.main.view_country_list_entry_flag_container.view.*
 import java.text.Collator
 
 class CountryList(context: Context, attrs: AttributeSet) :
     LinearLayout(context, attrs) {
-
+    private var attributeSet = attrs
     private var _list: List<Country>? = null
     var list: List<Country>?
         get() = _list
         set(value) {
             _list = value
-            buildList()
+            buildAllLists()
         }
 
     init {
-        orientation = VERTICAL
+        orientation = HORIZONTAL
+        inflate(context, R.layout.view_country_list_entry_flag_container, this)
     }
 
     /**
      * Cleans the view and rebuilds the list of countries. Presets already selected countries
      */
-    private fun buildList() {
-        this.removeAllViews()
+    private fun buildAllLists() {
         list
             ?.map { country ->
                 context.getString(country.labelRes) to country.iconRes
             }
             ?.sortedWith { a, b ->
                 Collator.getInstance().compare(a.first, b.first)
             }
-            ?.forEachIndexed { index, (label, iconRes) ->
-                inflate(context, R.layout.view_country_list_entry, this)
-                val child = this.getChildAt(index)
-                child.apply {
-                    findViewById<ImageView>(R.id.country_list_entry_image).setImageResource(iconRes)
-                    findViewById<TextView>(R.id.country_list_entry_label).text = label
-                }
+        list?.let { prepareFlags(it) }
+    }
+
+    private fun prepareFlags(list: List<Country>) {
+        val adapterCountryFlags = _list?.let { CountryAdapterFlags(it) }
+        var countryNames = ""
+        context.withStyledAttributes(attributeSet, R.styleable.SimpleStepEntry) {
+            flagGrid.layoutManager = GridLayoutManager(context, FLAG_COLUMNS)
+            flagGrid.adapter = adapterCountryFlags
+            val iterator: Iterator<Country> = list.iterator()
+            while (iterator.hasNext()) {
+                countryNames += resources.getString(iterator.next().labelRes)
+                if (iterator.hasNext()) countryNames += ", "
             }
+            country_list_entry_label.text = countryNames
+        }
+    }
+
+    companion object {
+        const val FLAG_COLUMNS = 8
+    }
+}
+
+/**
+ * Adapter for the country flags grid
+ */
+class CountryAdapterFlags internal constructor(
+    countryList: List<Country>
+) :
+    BaseCountryFlagAdapter<CountryAdapterFlags.CountryFlagViewHolder>() {
+    // total number of cells
+
+    private val countryList: List<Country>? = countryList
```

Comment:
I think the country list does not change for the adapter.

Try if we can just use a constructor property.

```kotlin
constructor(
    private val countryList: List<Country>
)
```